### DISCUSSION TITLE
relation access

### DISCUSSION THREAD

**Discussion by epipheus** [SeaQL/sea-orm#1044] [Q&A] [answered]
I have a very basic question, I must be missing something. Coming from RubyonRails/ActiveRecord experience where relations mean you have an accessor attached to an object by the name of the model. If you found a cake by id with `cake1 = Cake.find(some_id)` in you could access the filling with `cake1.filling` . In fact there is an accessor on both sides of a relation.  

In the Relations section of the documentation I really don't see an example of actually accessing the relation.  I am trying to understand how they are accessed. What am I missing?

**Comment by billy1624** [comment, marked as answer]
Hey @epipheus, welcome! Yes you could, similar to your example above. Let say we have a cake and we want to get the fruit (topping) of it.

```rs
// Find a cake model first
let cheese: Option<cake::Model> = Cake::find_by_id(1).one(db).await?;
let cheese: cake::Model = cheese.unwrap();

// Then, find all related fruits of this cake
let fruits: Vec<fruit::Model> = cheese.find_related(Fruit).all(db).await?;
```

Docs: https://www.sea-ql.org/SeaORM/docs/basic-crud/select/#lazy-loading

**Reply by epipheus** [reply]
This is helpful.  Feels verbose to need to manually load every relation when I have Entity machinery for reading tho

**Reply by billy1624** [reply]
You could also load it eagerly at once
```rs
let cake_with_fruits: Vec<(cake::Model, Vec<fruit::Model>)> = Cake::find()
    .find_with_related(Fruit)
    .all(db)
    .await?;
```
https://www.sea-ql.org/SeaORM/docs/basic-crud/select/#eager-loading

**Reply by numfin** [reply]
@billy1624 hi! What i have to do if i want to load fruits on topings of all cakes?

We load cakes `Cake::find().all()`
We load topings `.find_with_related()`
But how do i get fruits? Do i have to send multiple database requests just to make another join?

**Reply by frederikhors** [reply]
> @billy1624 hi! What i have to do if i want to load fruits on topings of all cakes?
> 
> We load cakes `Cake::find().all()` We load topings `.find_with_related()` But how do i get fruits? Do i have to send multiple database requests just to make another join?

This is the only issue we have with SeaORM right now and that is blocking the switch to it.

**Reply by numfin** [reply]
@frederikhors i found workaround :) https://github.com/SeaQL/sea-orm/issues/1465
Would you give me feedback if it worked for you?

**Comment by tyt2y3** [comment]
Thanks Billy. My few cents:

1. https://www.sea-ql.org/blog/2022-05-14-faq-01/ this could provide some context
2. Not an expert in RoR, but the 'accessor' you are looking for are the trait methods `find_related`. It does not automatically exist on the other side, but if you use the code generator, then it also generates the `impl Related` for the other Entity
3. Essentially the constraint in Rust is that 1) object attributes are not dynamic 2) attributes cannot be inherited as well. This is a good thing because it imposes a type system to your code, but if you are coming from a dynamic language certain kind of tricks are not doable in Rust

**Reply by epipheus** [reply]
This gets to the underlying heart of what I'm confused about and it seems it's all by design? but, it all feels insanely verbose! I guess I understand the tradeoff and I guess I assume that the purpose of an ORM is to make accessing the data easy.  I would bet that not taking the data ease path may ultimately hurts because of multiple queries. Verbosity of the code definitely hurts.  Manually loading every relation feels insane an definitely like something that should happen behind the scenes that I could override to eek out performance.

**Reply by frederikhors** [reply]
@epipheus I'm in your situation. I just opened https://github.com/SeaQL/sea-orm/discussions/1139 and found out this discussion.

Have you found a solution?

**Comment by epipheus** [comment]
I appreciate you both. i'm new to rust and sea-orm and everything feels verbose to me at this point.
